### DISCUSSION TITLE
Difference between pcp_combine -sum and -add

### DISCUSSION THREAD

**Discussion by BinbinZhou-NOAA** [dtcenter/METplus#1803] [Configuration] [answered]
What is the difference between pcp_combine's  -sum and -add options 
For only 2 input files, they are same?

**Comment by JohnHalleyGotway** [comment, marked as answer]
@BinbinZhou-NOAA, my answer to your question is *kind of*, but let me explain. While its true that running `-add` for two input files will produce the exact same numeric results as running `-sum` to locate and add them, the command line arguments are very different.

The `-sum` option uses:
```
	SUM_ARGS:
		init_time
		in_accum
		valid_time
		out_accum
```
While the `-add` option uses:
```
	INPUT_FILES:
		file_1 config_str_1 ... file_n config_str_n
```

For some background, the `-sum` option was written first many years ago. It only works with input data with a consistent accumulation interval, such as 3-hour accumulations of precipitation. It uses the command line arguments to decide what data to look for... dividing the requested `out_accum` by the `in_accum` to figure out the number of files to find. I searches through the input `-pcpdir` directory(ies), processing each file it finds that matches the `-pcprx` regular expression, looking for precipitation (APCP) data that matches the one of the timesteps.

Whereas the newer `-add` option is much simpler. It just reads the requested data from the files specified on the command line.

The following 2 commands produce the same result:
```
pcp_combine -sum 20050807_000000 3 20050807_060000 06 sum_06a.nc -pcpdir data/sample_fcst/2005080700
```
and
```
pcp_combine -add data/sample_fcst/2005080700/wrfprs_ruc13_03.tm00_G212 data/sample_fcst/2005080700/wrfprs_ruc13_06.tm00_G212 -field 'name="APCP"; level="A03";' add_06a.nc
```

The `-sum` option is limited to only processing "accumulating" variables, like APCP. And I think it only works well with GRIB inputs. It's also slow because it has to go open up many files looking for the matching data.

The `-add` option is faster since it only opens the exact files that it needs. The harder part is figuring out which files to pass to it... but that's the advantage of METplus. The METplus wrapper will build the call to pcp_combine for you, as long as you've configured it well.

**Reply by BinbinZhou-NOAA** [reply]
I also observed the the -sum option is very slow.  Based on your explanation, -add can replace -sum option. 
and will run faster.

Thanks!

**Reply by JohnHalleyGotway** [reply]
@BinbinZhou-NOAA, great, glad that explanation help. Yes, I would also advise using `-add` instead of `-sum` for the reasons you've already found. I'll go ahead and mark this discussion as being answered and also "lock it" to prevent future posts.

As always, feel free to create new discussions when new issues or questions arise.
